pkg/platform/model/buildplanner.go

```diff
@@ -600,3 +601,98 @@ func VersionStringToRequirements(version string) ([]bpModel.VersionRequirement,
 	}
 	return requirements, nil
 }
+
+func (bp *BuildPlanner) BuildTarget(owner, project, commitID, target string) error {
+	logging.Debug("BuildTarget, owner: %s, project: %s, commitID: %s, target: %s", owner, project, commitID, target)
+	resp := &bpModel.BuildTargetResult{}
+	err := bp.client.Run(request.Evaluate(owner, project, commitID, target), resp)
+	if err != nil {
+		return processBuildPlannerError(err, "Failed to evaluate target")
+	}
+
+	if resp.Project == nil {
+		return errs.New("Project is nil")
+	}
+
+	if bpModel.IsErrorResponse(resp.Project.Type) {
+		return bpModel.ProcessProjectError(resp.Project, "Could not evaluate target")
+	}
+
+	if resp.Project.Commit == nil {
+		return errs.New("Commit is nil")
+	}
+
+	if bpModel.IsErrorResponse(resp.Project.Commit.Type) {
+		return bpModel.ProcessCommitError(resp.Project.Commit, "Could not process error response from evaluate target")
+	}
+
+	if resp.Project.Commit.Build == nil {
+		return errs.New("Build is nil")
+	}
+
+	if bpModel.IsErrorResponse(resp.Project.Commit.Build.Type) {
+		return bpModel.ProcessBuildError(resp.Project.Commit.Build, "Could not process error response from evaluate target")
+	}
+
+	if err := bp.pollBuildStatus(commitID); err != nil {
+		return errs.Wrap(err, "Failed to poll build status")
+	}
+
+	return nil
+}
+
+func (bp *BuildPlanner) pollBuildStatus(commitID string) error {
+	resp := model.NewBuildPlanResponse("", "")
+	ticker := time.NewTicker(pollInterval)
+	for {
+		select {
+		case <-ticker.C:
+			err := bp.client.Run(request.BuildPlan(commitID, "", ""), resp)
+			if err != nil {
+				return processBuildPlannerError(err, "failed to fetch build plan")
+			}
+
+			if resp == nil {
+				return errs.New("Build plan response is nil")
+			}
+
+			build, err := resp.Build()
+			if err != nil {
+				return errs.Wrap(err, "Could not get build from response")
+			}
+
+			// The type aliasing in the query populates the
+			// response with emtpy targets that we should remove
+			removeEmptyTargets(build)
+
+			// If the build status is planning it may not have any artifacts yet.
+			if build.Status != bpModel.Planning {
+				continue
+			}
+
+			// If all artifacts are completed then we are done.
+			completed := true
+			for _, artifact := range build.Artifacts {
+				if artifact.Status == bpModel.ArtifactFailedPermanently ||
+					artifact.Status == bpModel.ArtifactFailedTransiently {
+					return errs.New("Artifact %s failed", artifact.NodeID)
+				}
+
+				if artifact.Status == bpModel.ArtifactNotSubmitted {
+					continue
+				}
+
+				if artifact.Status != bpModel.ArtifactSucceeded {
+					completed = false
+					break
+				}
+			}
+
+			if completed {
+				return nil
+			}
```

Comment:
We should also have a fallback for `build.Status == bpModel.Completed` here imo. Once we hit that there is no point to keep polling, but the code as written could technically do this.

Reply:
Unfortunately, the build plan status does not accurately reflect the overall build. There is a story to get this updated but for now the solution we decided on was to check each artifact status.

See my question [here](https://activestate.slack.com/archives/C03T15HMBMG/p1709743461056899) and the follow-up story [here](https://activestatef.atlassian.net/browse/AR-127)

Reply:
Yeah but I think that only speaks to the false negatives, not false positives, right?

In other words; it might not say the status is completed, even though it is. But if it says it is completed, it is for sure done?

Reply:
Ah I see. Yes, you're correct. Looking at the docstring for the build plan statuses `COMPLETED` should work here.